File: samples/python/workflow_overview.py
"""Single-file reference implementation of the Make connection workflow.

This script mirrors the behaviour of the richer Node service but keeps everything
in one place so developers can port the logic to other stacks with minimal effort.
"""
from __future__ import annotations

import json
import os
import sys
import webbrowser
from pathlib import Path
from typing import Dict, Iterable, List, Optional

import requests

ROOT_DIR = Path(__file__).resolve().parents[2]
ENV_PATH = ROOT_DIR / ".env"
DEFAULT_BLUEPRINT_DIR = ROOT_DIR / "blueprints"
# We drop updated blueprints next to the sample so each language stays self-contained.
UPDATED_DIR = Path(__file__).resolve().parent


def load_env(path: Path) -> Dict[str, str]:
    """Load .env style key=value pairs without introducing a new dependency."""
    values: Dict[str, str] = {}
    if not path.exists():
        return values
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        values[key.strip()] = value.strip().strip('"')
    return values


def ensure_settings(env: Dict[str, str]) -> None:
    """Fail fast if the caller forgot to populate the Make API essentials."""
    missing = [key for key in ("INSTANCE_URL", "AUTH_TOKEN", "TEAM_ID") if not env.get(key)]
    if missing:
        raise RuntimeError(f"Missing required env vars: {', '.join(missing)}")


def make_headers(env: Dict[str, str]) -> Dict[str, str]:
    """Build the baseline Make API headers once so wrappers can stay tiny."""
    return {
        "Content-Type": "application/json",
        "Authorization": f"Token {env['AUTH_TOKEN']}",
    }


def fetch_json(env: Dict[str, str], path: str, *, method: str = "GET", body: Optional[dict] = None) -> dict:
    """Small helper for JSON Make API calls (GET by default, POST when body provided)."""
    url = f"{env['INSTANCE_URL'].rstrip('/')}{path}"
    response = requests.request(method, url, headers=make_headers(env), data=json.dumps(body) if body else None)
    response.raise_for_status()
    return response.json()


def fetch_raw(env: Dict[str, str], path: str, *, method: str = "GET", allow_redirects: bool = True) -> requests.Response:
    """Variant of fetch_json that exposes the raw response (used for consent redirects)."""
    url = f"{env['INSTANCE_URL'].rstrip('/')}{path}"
    response = requests.request(method, url, headers=make_headers(env), allow_redirects=allow_redirects)
    response.raise_for_status()
    return response


def load_blueprint(name_or_path: str) -> dict:
    """Open a blueprint either by absolute path or by name inside the default folder."""
    candidate_path = Path(name_or_path)
    if not candidate_path.exists():
        candidate_path = DEFAULT_BLUEPRINT_DIR / name_or_path
    if not candidate_path.exists():
        raise FileNotFoundError(f"Blueprint not found: {name_or_path}")
    return json.loads(candidate_path.read_text())


def analyse_blueprint(blueprint: dict) -> tuple[str, List[str]]:
    """Walk through the flow and return the non-Facebook app plus its module names."""
    modules: List[tuple[str, str]] = []

    def walk(node):
        if not node:
            return
        if isinstance(node, list):
            for item in node:
                walk(item)
            return
        if isinstance(node, dict):
            module = node.get("module")
            if isinstance(module, str):
                app, _, action = module.partition(":")
                if app and action:
                    modules.append((app, action))
            for value in node.values():
                walk(value)

    walk(blueprint.get("flow"))
    filtered = [item for item in modules if item[0] != "facebook-conversion-leads"]
    if not filtered:
        raise RuntimeError("No eligible modules found in blueprint")

    app_name = filtered[0][0]
    module_names = [action for app, action in filtered if app == app_name]
    return app_name, module_names


def collect_scopes(app_definition: dict, module_names: Iterable[str]) -> List[str]:
    """Cross-reference module names against the app definition to gather scopes."""
    lookup: Dict[str, List[str]] = {}
    for bucket in ("actions", "searches", "triggers"):
        for item in app_definition.get("app", {}).get(bucket, []) or []:
            name = item.get("name")
            if name:
                scopes = item.get("scopes") or item.get("scope") or []
                lookup[name.lower()] = scopes

    combined: set[str] = set()
    for name in module_names:
        combined.update(lookup.get(name.lower(), []))
    return sorted(combined)


def collect_form_fields(schema: dict, bucket: Optional[List[dict]] = None) -> List[dict]:
    """Flatten the form schema and keep only user-editable fields we must populate."""
    bucket = bucket or []
    if not schema:
        return bucket
    if isinstance(schema, list):
        for item in schema:
            collect_form_fields(item, bucket)
        return bucket
    if isinstance(schema, dict):
        identifier = schema.get("key") or schema.get("name")
        field_type = schema.get("type") or schema.get("templateOptions", {}).get("type")
        if identifier and (field_type or "").lower() not in {"button", "content", "htmlelement"}:
            cloned = {
                "key": identifier,
                "type": field_type,
                "data": schema.get("data"),
                "templateOptions": schema.get("templateOptions"),
                "required": bool(
                    schema.get("required")
                    or schema.get("validate", {}).get("required")
                    or schema.get("templateOptions", {}).get("required")
                ),
            }
            template_options = schema.get("templateOptions") or {}
            default_value = schema.get("defaultValue")
            if default_value is None:
                default_value = template_options.get("defaultValue")
            if default_value is None:
                default_value = template_options.get("default")
            if default_value is None:
                default_value = schema.get("default")
            if default_value is not None:
                cloned["defaultValue"] = default_value
            bucket.append(cloned)

        for child_key, child_value in schema.items():
            if child_key == "components":
                continue
            if child_key == "options" and isinstance(child_value, dict):
                store = child_value.get("store")
                if isinstance(store, list) and store:
                    selected = next(
                        (
                            option
                            for option in store
                            if isinstance(option, dict) and (option.get("default") is True or option.get("selected") is True)
                        ),
                        None,
                    )
                    if selected is None:
                        selected = store[0]
                    if isinstance(selected, dict):
                        collect_form_fields(selected, bucket)
                continue
            if child_key == "nested" and isinstance(child_value, list):
                for item in child_value:
                    collect_form_fields(item, bucket)
                continue
            collect_form_fields(child_value, bucket)
    return bucket


def build_payload(form_schema: dict, app_name: str, scopes: List[str], account_name_override: Optional[str]) -> dict:
    """Merge defaults from the form schema with our derived scopes and account name."""
    payload: Dict[str, object] = {}
    fields = collect_form_fields(form_schema)
    for field in fields:
        key = field.get("key")
        if not key:
            continue
        if "defaultValue" in field:
            payload[key] = field.get("defaultValue")
        else:
            payload[key] = field_fallback(field)

    if isinstance(payload.get("accountName"), str) and "{{" in payload["accountName"] and "}}" in payload["accountName"]:
        payload["accountName"] = ""

    payload["accountType"] = payload.get("accountType") or app_name
    payload["accountName"] = account_name_override or payload.get("accountName") or f"{app_name} connection"
    payload["customScopes"] = scopes

    if "property" in payload and not payload["property"]:
        raise RuntimeError("The connection form expects a property value")
    return payload


def field_fallback(field: dict) -> object:
    field_type = str(field.get("type") or "").lower()
    template_type = str(field.get("templateOptions", {}).get("type") or "").lower()
    if field_type == "boolean" or template_type == "boolean":
        return False
    candidate = option_fallback(field)
    if candidate is not None:
        return candidate
    return ""


def option_fallback(field: dict) -> Optional[object]:
    data = field.get("data")
    template = field.get("templateOptions")
    sources = [source for source in (data, template) if isinstance(source, dict)]
    if not sources:
        return None

    for source in sources:
        if "defaultValue" in source:
            return source.get("defaultValue")

    for source in sources:
        for key in ("options", "values", "items", "enum"):
            options = source.get(key)
            if isinstance(options, dict) and isinstance(options.get("store"), list):
                options = options.get("store")
            if not isinstance(options, list) or not options:
                continue
            preferred = next((opt for opt in options if isinstance(opt, dict) and opt.get("default") is True), None)
            if preferred is None:
                preferred = next((opt for opt in options if isinstance(opt, dict) and opt.get("selected") is True), None)
            candidates = options if preferred is None else [preferred]
            for option in candidates:
                if isinstance(option, dict):
                    if "value" in option:
                        return option["value"]
                    if "id" in option:
                        return option["id"]
                    if "code" in option:
                        return option["code"]
                elif option is not None:
                    return option
    return None


def save_updated_blueprint(original_name: str, blueprint: dict, connection_id: int) -> Path:
    """Write a copy of the source blueprint with fresh connection IDs."""
    clone = json.loads(json.dumps(blueprint))
    updated = 0

    def rewrite(node):
        nonlocal updated
        if not node:
            return
        if isinstance(node, list):
            for item in node:
                rewrite(item)
            return
        if isinstance(node, dict):
            module = node.get("module")
            if isinstance(module, str) and not module.startswith("facebook-conversion-leads"):
                parameters = node.get("parameters", {})
                if "__IMTCONN__" in parameters:
                    parameters["__IMTCONN__"] = connection_id
                    updated += 1
            for value in node.values():
                rewrite(value)

    rewrite(clone.get("flow"))

    UPDATED_DIR.mkdir(parents=True, exist_ok=True)
    target_path = UPDATED_DIR / f"{Path(original_name).stem}-updated.json"
    target_path.write_text(json.dumps(clone, indent=4) + "\n")
    print(f"Updated {updated} module(s) in {target_path}")
    return target_path


def run_workflow(env: Dict[str, str], blueprint_name: str, account_name: Optional[str]) -> dict:
    """End-to-end orchestration: analyse blueprint, create connection, request consent."""
    blueprint = load_blueprint(blueprint_name)
    app_name, module_names = analyse_blueprint(blueprint)

    app_definition = fetch_json(env, f"/api/v2/imt/apps/{app_name}")
    scopes = collect_scopes(app_definition, module_names)
    form_schema = fetch_json(env, f"/api/v2/imt-forms/connections/create?type={app_name}&teamId={env['TEAM_ID']}")
    payload = build_payload(form_schema, app_name, scopes, account_name)

    response = fetch_json(env, f"/api/v2/connections?teamId={env['TEAM_ID']}&inspector=0", method="POST", body=payload)
    connection = response.get("connection") if isinstance(response, dict) else None
    if connection is None and isinstance(response, dict):
        connection = response
    if not isinstance(connection, dict) or "id" not in connection:
        raise RuntimeError(f"Connection response missing id: {json.dumps(response, indent=2)}")
    consent_response = fetch_raw(env, f"/api/v2/oauth/auth/{connection['id']}", allow_redirects=False)
    consent_url = consent_response.headers["Location"]

    updated_path = save_updated_blueprint(blueprint_name, blueprint, connection["id"])

    return {
        "connection": connection,
        "consent_url": consent_url,
        "payload": payload,
        "updated_blueprint": updated_path,
    }


def main(argv: List[str]) -> int:
    """CLI entry point for parity with the other sample languages."""
    if len(argv) < 2:
        print("Usage: python workflow_overview.py <blueprint.json> [account-name]", file=sys.stderr)
        return 1

    blueprint_name = argv[1]
    account_name = argv[2] if len(argv) > 2 else None

    env = {**load_env(ENV_PATH), **os.environ}
    ensure_settings(env)

    result = run_workflow(env, blueprint_name, account_name)
    print(f"Connection created: {result['connection']['id']}")
    print(f"Consent URL: {result['consent_url']}")
    print(f"Updated blueprint written to: {result['updated_blueprint']}")

    try:
        webbrowser.open(result["consent_url"], new=2)
        print("Consent page opened in your browser.")
    except Exception as exc:  # noqa: broad-except
        print(f"Unable to open browser automatically: {exc}")

    input("Complete the consent flow, then press Enter to trigger the connection test...")
    test_result = fetch_json(env, f"/api/v2/connections/{result['connection']['id']}/test", method="POST")
    print("Test response:")
    print(json.dumps(test_result, indent=2))

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: samples/rust/workflow_overview.rs
//! Simplified Make connection workflow in Rust.
//!
//! This program mirrors the single-file examples in Node/Python/Go. It uses
//! `reqwest` and `serde_json` so you can compile it quickly and use the code as
//! a translation guide.
//!
//! Usage:
//!     cargo run --example workflow_overview "Hubspot - FBcapi.json" "Optional Account Name"
use std::collections::{BTreeMap, HashMap, HashSet};
use std::env;
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

use reqwest::blocking::{Client, Response};
use reqwest::header::{HeaderMap, HeaderValue, CONTENT_TYPE};
use serde_json::{json, Value};

fn main() -> anyhow::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run --example workflow_overview <blueprint.json> [account-name]");
        std::process::exit(1);
    }

    let blueprint_name = &args[1];
    let account_override = args.get(2).cloned();

    let root = workspace_root()?;
    let env_path = root.join(".env");
    let env = load_env(&env_path)?;
    ensure_settings(&env)?;

    let client = Client::new();
    let blueprint = load_blueprint(&root, blueprint_name)?;
    let (app_name, modules) = analyse_blueprint(&blueprint);

    let app_definition = fetch_json(&client, &env, &format!("/api/v2/imt/apps/{app_name}"))?;
    let scopes = collect_scopes(&app_definition, &modules);
    let form_schema = fetch_json(&client, &env, &format!(
        "/api/v2/imt-forms/connections/create?type={app_name}&teamId={}",
        env["TEAM_ID"]
    ))?;
    let payload = build_payload(&form_schema, &app_name, &scopes, account_override.as_deref());

    let create_resp = fetch_json_with_body(
        &client,
        &env,
        &format!("/api/v2/connections?teamId={}&inspector=0", env["TEAM_ID"]),
        payload.clone(),
    )?;
    let connection_value = create_resp
        .get("connection")
        .cloned()
        .unwrap_or_else(|| create_resp.clone());
    let connection_obj = connection_value
        .as_object()
        .ok_or_else(|| anyhow::anyhow!("Connection response missing object payload: {}", connection_value))?;
    let connection_id = connection_obj
        .get("id")
        .cloned()
        .ok_or_else(|| anyhow::anyhow!("Connection response missing id: {}", connection_value))?;

    let consent_url = fetch_consent_url(&client, &env, connection_id.clone())?;
    let updated_path = save_blueprint(&root.join("samples/rust"), blueprint_name, &blueprint, &connection_id)?;

    println!("Connection created: {}", connection_id);
    println!("Consent URL: {}", consent_url);
    println!("Updated blueprint written to: {}", updated_path.display());

    if webbrowser::open(&consent_url).is_ok() {
        println!("Consent page opened in your browser.");
    }

    print!("Complete consent, then press Enter to run the connection test...");
    io::stdout().flush().ok();
    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer).ok();

    let test_resp = fetch_json_with_body(
        &client,
        &env,
        &format!("/api/v2/connections/{}/test", connection_id),
        Value::Null,
    )?;
    println!("Test response:\n{}", serde_json::to_string_pretty(&test_resp)?);

    Ok(())
}

/// Locate the repository root so the script mirrors the behaviour of the other samples.
fn workspace_root() -> anyhow::Result<PathBuf> {
    let exe = env::current_exe()?;
    Ok(exe.parent().unwrap().join("..").join(".."))
}

/// Read a minimal `.env` file without pulling in an external parser crate.
fn load_env(path: &Path) -> anyhow::Result<HashMap<String, String>> {
    let mut values = HashMap::new();
    if !path.exists() {
        return Ok(values);
    }
    for line in fs::read_to_string(path)?.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        if let Some((key, value)) = trimmed.split_once('=') {
            values.insert(key.trim().to_owned(), value.trim_matches('"').to_owned());
        }
    }
    Ok(values)
}

/// Stop early if the core Make settings are missing instead of failing mid-request.
fn ensure_settings(env: &HashMap<String, String>) -> anyhow::Result<()> {
    let mut missing = Vec::new();
    for key in ["INSTANCE_URL", "AUTH_TOKEN", "TEAM_ID"] {
        if env.get(key).map(|s| s.is_empty()).unwrap_or(true) {
            missing.push(key);
        }
    }
    if !missing.is_empty() {
        anyhow::bail!("Missing required env vars: {}", missing.join(", "));
    }
    Ok(())
}

/// Load a blueprint either via absolute path or by name from the standard folder.
fn load_blueprint(root: &Path, name: &str) -> anyhow::Result<Value> {
    let candidates = [Path::new(name).to_path_buf(), root.join("blueprints").join(name)];
    for candidate in &candidates {
        if candidate.exists() {
            let text = fs::read_to_string(candidate)?;
            return Ok(serde_json::from_str(&text)?);
        }
    }
    anyhow::bail!("Blueprint not found: {name}");
}

/// Return the first non-Facebook app and all module names tied to it.
fn analyse_blueprint(blueprint: &Value) -> (String, Vec<String>) {
    let mut modules = Vec::<(String, String)>::new();

    fn walk(node: &Value, acc: &mut Vec<(String, String)>) {
        match node {
            Value::Array(items) => items.iter().for_each(|item| walk(item, acc)),
            Value::Object(map) => {
                if let Some(Value::String(module)) = map.get("module") {
                    if let Some((app, action)) = module.split_once(':') {
                        acc.push((app.to_string(), action.to_string()));
                    }
                }
                map.values().for_each(|value| walk(value, acc));
            }
            _ => {}
        }
    }

    walk(&blueprint["flow"], &mut modules);
    for (app, _) in &modules {
        if app != "facebook-conversion-leads" {
            let actions = modules
                .iter()
                .filter(|(candidate, _)| candidate == app)
                .map(|(_, action)| action.clone())
                .collect();
            return (app.clone(), actions);
        }
    }
    panic!("No eligible modules found");
}

/// Cross-reference module names against the app definition to gather required scopes.
fn collect_scopes(app_definition: &Value, module_names: &[String]) -> Vec<String> {
    let mut lookup: HashMap<String, Vec<String>> = HashMap::new();
    if let Some(app) = app_definition.get("app") {
        for bucket in ["actions", "searches", "triggers"] {
            if let Some(Value::Array(items)) = app.get(bucket) {
                for item in items {
                    if let Some(Value::String(name)) = item.get("name") {
                        if let Some(scopes) = item.get("scopes") {
                            lookup.insert(name.to_lowercase(), to_string_vec(scopes));
                        }
                    }
                }
            }
        }
    }
    let mut combined = HashSet::new();
    for name in module_names {
        if let Some(scopes) = lookup.get(&name.to_lowercase()) {
            combined.extend(scopes.iter().cloned());
        }
    }
    combined.into_iter().collect()
}

/// Combine default form fields, the computed scopes, and the optional account override.
fn build_payload(
    schema: &Value,
    app_name: &str,
    scopes: &[String],
    account_override: Option<&str>,
) -> Value {
    let mut payload = BTreeMap::<String, Value>::new();
    let mut fields: Vec<serde_json::Map<String, Value>> = Vec::new();

    fn collect(node: &Value, bucket: &mut Vec<serde_json::Map<String, Value>>) {
        match node {
            Value::Array(items) => items.iter().for_each(|item| collect(item, bucket)),
            Value::Object(map) => {
                let identifier = map
                    .get("key")
                    .and_then(Value::as_str)
                    .map(|s| s.to_string())
                    .or_else(|| map.get("name").and_then(Value::as_str).map(|s| s.to_string()));

                let resolved_type = map
                    .get("type")
                    .and_then(Value::as_str)
                    .map(|s| s.to_string())
                    .or_else(|| map
                        .get("templateOptions")
                        .and_then(|value| value.get("type"))
                        .and_then(Value::as_str)
                        .map(|s| s.to_string()));

                let lower_type = resolved_type.as_deref().unwrap_or("").to_lowercase();
                if let Some(key) = identifier {
                    if !matches!(lower_type.as_str(), "button" | "content" | "htmlelement") {
                        let mut field = serde_json::Map::new();
                        field.insert("key".into(), Value::String(key));
                        if let Some(t) = resolved_type {
                            field.insert("type".into(), Value::String(t));
                        }
                        if let Some(data) = map.get("data") {
                            field.insert("data".into(), data.clone());
                        }
                        if let Some(template) = map.get("templateOptions") {
                            field.insert("templateOptions".into(), template.clone());
                        }

                        let mut default_value = map.get("defaultValue").cloned();
                        if default_value.is_none() {
                            default_value = map
                                .get("templateOptions")
                                .and_then(|value| value.get("defaultValue").cloned())
                                .or_else(|| map
                                    .get("templateOptions")
                                    .and_then(|value| value.get("default").cloned()))
                                .or_else(|| map.get("default").cloned());
                        }
                        if let Some(value) = default_value {
                            field.insert("defaultValue".into(), value);
                        }

                        bucket.push(field);
                    }
                }

                for (prop_key, value) in map {
                    match prop_key.as_str() {
                        "components" => {}
                        "options" => {
                            if let Value::Object(inner) = value {
                                if let Some(Value::Array(store)) = inner.get("store") {
                                    let mut selected: Option<&Value> = None;
                                    for option in store {
                                        if let Some(default_flag) = option.get("default").and_then(Value::as_bool) {
                                            if default_flag {
                                                selected = Some(option);
                                                break;
                                            }
                                        }
                                        if let Some(selected_flag) = option.get("selected").and_then(Value::as_bool) {
                                            if selected_flag {
                                                selected = Some(option);
                                                break;
                                            }
                                        }
                                        if selected.is_none() {
                                            selected = Some(option);
                                        }
                                    }
                                    if let Some(option) = selected {
                                        collect(option, bucket);
                                    }
                                    continue;
                                }
                            }
                            collect(value, bucket);
                        }
                        "nested" => {
                            if let Value::Array(items) = value {
                                for entry in items {
                                    collect(entry, bucket);
                                }
                            }
                        }
                        _ => collect(value, bucket),
                    }
                }
            }
            _ => {}
        }
    }

    collect(schema, &mut fields);


    for field in fields {
        if let Some(Value::String(key)) = field.get("key") {
            if let Some(value) = field.get("defaultValue") {
                payload.insert(key.clone(), value.clone());
            } else {
                payload.insert(key.clone(), fallback_for_field(&field));
            }
        }
    }

    if let Some(Value::String(name)) = payload.get("accountName") {
        if name.contains("{{") && name.contains("}}") {
            payload.insert("accountName".into(), Value::String(String::new()));
        }
    }

    payload.entry("accountType".into()).or_insert(Value::String(app_name.to_string()));
    if let Some(override_name) = account_override {
        payload.insert("accountName".into(), Value::String(override_name.to_string()));
    } else {
        payload.entry("accountName".into()).or_insert(Value::String(format!("{app_name} connection")));
    }
    payload.insert("customScopes".into(), Value::Array(scopes.iter().cloned().map(Value::String).collect()));
    if let Some(Value::String(prop)) = payload.get("property") {
        if prop.is_empty() {
            panic!("The connection form expects a property value");
        }
    }
    Value::Object(payload)
}

fn fallback_for_field(map: &serde_json::Map<String, Value>) -> Value {
    let template_type = map
        .get("templateOptions")
        .and_then(|value| value.get("type"))
        .and_then(Value::as_str)
        .map(|s| s.eq_ignore_ascii_case("boolean"))
        .unwrap_or(false);
    if let Some(Value::String(field_type)) = map.get("type") {
        if field_type.eq_ignore_ascii_case("boolean") || template_type {
            return Value::Bool(false);
        }
    } else if template_type {
        return Value::Bool(false);
    }
    if let Some(candidate) = option_fallback(map) {
        return candidate;
    }
    Value::String(String::new())
}

fn option_fallback(map: &serde_json::Map<String, Value>) -> Option<Value> {
    let mut sources = Vec::new();
    if let Some(Value::Object(data)) = map.get("data") {
        sources.push(data);
    }
    if let Some(Value::Object(template)) = map.get("templateOptions") {
        sources.push(template);
    }
    if sources.is_empty() {
        return None;
    }

    for source in &sources {
        if let Some(value) = source.get("defaultValue") {
            return Some(value.clone());
        }
    }

    for source in sources {
        for key in ["options", "values", "items", "enum"] {
            let mut candidates: Vec<Value> = Vec::new();
            match source.get(key) {
                Some(Value::Array(items)) => candidates.extend(items.clone()),
                Some(Value::Object(map)) => {
                    if let Some(Value::Array(store)) = map.get("store") {
                        candidates.extend(store.clone());
                    }
                }
                _ => {}
            }
            let preferred = candidates.iter().find(|item| {
                item.get("default").and_then(Value::as_bool) == Some(true)
                    || item.get("selected").and_then(Value::as_bool) == Some(true)
            }).cloned();

            let iter: Vec<Value> = if let Some(pref) = preferred {
                vec![pref]
            } else {
                candidates.clone()
            };

            for item in iter {
                if let Value::Object(option) = item {
                    for candidate_key in ["value", "id", "code"] {
                        if let Some(value) = option.get(candidate_key) {
                            return Some(value.clone());
                        }
                    }
                } else if !item.is_null() {
                    return Some(item.clone());
                }
            }
        }
    }
    None
}

/// Clone the blueprint and swap `__IMTCONN__` for the new connection everywhere.
fn rewrite_blueprint(original: &Value, connection_id: &Value) -> (Value, usize) {
    let mut clone = original.clone();
    let mut updated = 0;

    fn walk(node: &mut Value, connection_id: &Value, updated: &mut usize) {
        match node {
            Value::Array(items) => items.iter_mut().for_each(|item| walk(item, connection_id, updated)),
            Value::Object(map) => {
                if let Some(Value::String(module)) = map.get("module") {
                    if !module.starts_with("facebook-conversion-leads") {
                        if let Some(Value::Object(params)) = map.get_mut("parameters") {
                            if params.contains_key("__IMTCONN__") {
                                params.insert("__IMTCONN__".into(), connection_id.clone());
                                *updated += 1;
                            }
                        }
                    }
                }
                map.values_mut().for_each(|value| walk(value, connection_id, updated));
            }
            _ => {}
        }
    }

    walk(&mut clone["flow"], connection_id, &mut updated);
    (clone, updated)
}

/// Persist the updated blueprint next to the sample so everything stays portable.
fn save_blueprint(dir: &Path, original_name: &str, blueprint: &Value, connection_id: &Value) -> anyhow::Result<PathBuf> {
    fs::create_dir_all(dir)?;
    let base = Path::new(original_name).file_stem().unwrap().to_string_lossy();
    let path = dir.join(format!("{base}-updated.json"));
    let (rewritten, updated) = rewrite_blueprint(blueprint, connection_id);
    println!("Updated {} module(s) in {}", updated, path.display());
    fs::write(&path, serde_json::to_string_pretty(&rewritten)? + "\n")?;
    Ok(path)
}

/// Convenience wrapper for GET requests that expect JSON bodies.
fn fetch_json(client: &Client, env: &HashMap<String, String>, path: &str) -> anyhow::Result<Value> {
    let response = fetch(client, env, path, reqwest::Method::GET, Value::Null, true)?;
    Ok(response.json()?)
}

/// POST helper that mirrors `fetch_json` but accepts an explicit JSON body.
fn fetch_json_with_body(
    client: &Client,
    env: &HashMap<String, String>,
    path: &str,
    body: Value,
) -> anyhow::Result<Value> {
    let response = fetch(client, env, path, reqwest::Method::POST, body, true)?;
    Ok(response.json()?)
}

/// Retrieve the consent redirect without following it to keep the browser in charge.
fn fetch_consent_url(client: &Client, env: &HashMap<String, String>, connection_id: Value) -> anyhow::Result<String> {
    let path = format!("/api/v2/oauth/auth/{}", connection_id);
    let response = fetch(client, env, &path, reqwest::Method::GET, Value::Null, false)?;
    if let Some(location) = response.headers().get("Location") {
        Ok(location.to_str()?.to_string())
    } else {
        anyhow::bail!("Missing Location header when requesting consent URL");
    }
}

/// Shared HTTP helper so all requests reuse the same error handling and headers.
fn fetch(
    client: &Client,
    env: &HashMap<String, String>,
    path: &str,
    method: reqwest::Method,
    body: Value,
    allow_redirects: bool,
) -> anyhow::Result<Response> {
    let url = format!("{}{}", env["INSTANCE_URL"].trim_end_matches('/'), path);
    let mut headers = HeaderMap::new();
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    headers.insert("Authorization", HeaderValue::from_str(&format!("Token {}", env["AUTH_TOKEN"]))?);

    let builder = client.request(method, &url).headers(headers);
    let builder = if body.is_null() { builder } else { builder.body(body.to_string()) };
    let client = if allow_redirects { client.clone() } else { Client::builder().redirect(reqwest::redirect::Policy::none()).build()? };
    let response = client.execute(builder.build()?)?;
    if response.status().is_success() || (!allow_redirects && response.status().is_redirection()) {
        Ok(response)
    } else {
        anyhow::bail!("Request failed: {}", response.status())
    }
}

/// Convert a JSON array into a vector of strings while dropping non-string entries.
fn to_string_vec(value: &Value) -> Vec<String> {
    value
        .as_array()
        .unwrap_or(&vec![])
        .iter()
        .filter_map(|item| item.as_str().map(|s| s.to_string()))
        .collect()
}

/// Mirrors the helper used in other languages; handy if you expand the sample further.
fn firstNonEmpty(values: &[String]) -> String {
    for value in values {
        if !value.trim().is_empty() {
            return value.clone();
        }
    }
    String::new()
}
